### PR TITLE
Add `strip_prefix` arg to bzlmod's `git_override`

### LINKED ISSUE #22076: Have MODULE's `git_override` support using a nested path
### Description of the feature request:

The [`git_override`](https://bazel.build/rules/lib/globals/module#git_override) MODULE.bazel method does not support pulling data from an external git repo's subdirectory.

This issue came up as I was trying to use a fork of `rules_python`'s `rules_python_gazelle_plugin` bazel module, which lives in a subdirectory of the repo https://github.com/bazelbuild/rules_python/blob/main/gazelle/MODULE.bazel.

```starlark
git_override(
    module_name = "rules_python_gazelle_plugin",
    remote = "https://github.com/dougthor42/rules_python",
    commit = "e1677949eca2abe4c37e5f5b2a69529deb44a3d4",
    use_subdir = "gazelle",  # or something.
)
```

After chatting with @Wyverald about it, it looks like the feature request is to add support for `strip_prefex` to `git_override`. However, I haven't fully groked how that would work.

### Which category does this issue belong to?

Configurability, Core

### What underlying problem are you trying to solve with this feature?

[`rules_python`](https://github.com/bazelbuild/rules_python)'s gazelle plugin resides in a [subdirectory of the `rules_python` repo](https://github.com/bazelbuild/rules_python/blob/main/gazelle/).

A naive first attempt at using `git_override` looks like so:

```starlark
# MODULE.bazel
bazel_dep(name = "rules_python", version = "0.31.0")
bazel_dep(name = "gazelle", version = "0.35.0", repo_name = "bazel_gazelle")
bazel_dep(name = "rules_python_gazelle_plugin", version = "0.31.0")

git_override(
    module_name = "rules_python",
    remote = "https://github.com/dougthor42/rules_python",
    commit = "e1677949eca2abe4c37e5f5b2a69529deb44a3d4",
)
```

However, this does not update `rules_python_gazelle_plugin`. So we try adding that too:

```starlark
git_override(
    module_name = "rules_python_gazelle_plugin",
    remote = "https://github.com/dougthor42/rules_python",
    commit = "e1677949eca2abe4c37e5f5b2a69529deb44a3d4",
)
```

But bazel rightly complains that the `module_name` provided does not match the module name of the git repo.

So lastly we try to point to the `gazelle` subdir:

```starlark
git_override(
    module_name = "rules_python_gazelle_plugin",
    remote = "https://github.com/dougthor42/rules_python/gazelle",
    commit = "e1677949eca2abe4c37e5f5b2a69529deb44a3d4",
)
```

Obviously, this fails because `https://github.com/dougthor42/rules_python/gazelle` does not exist.

### Which operating system are you running Bazel on?

gLinux

### What is the output of `bazel info release`?

release 7.1.1

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

N/A

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?

```text
N/A
```

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

### PR DESCRIPTION
**Pull request by dougthor42**
Add `strip_prefix` arg go bzlmod's `git_override` and pass it down to `git_repository`.

This allows users to work with git repositories that have other Bazel modules in subdirectories.
A prime example of such is the `rules_python`](https://github.com/bazelbuild/rules_python) repo:

```
./rules_python
├── MODULE.bazel       # available on BCR as "rules_python"
└── gazelle
    └── MODULE.bazel   # available on BCR as "rules_python_gazelle_plugin"
```

Fixes #22076.

RELNOTES:
bzlmod `git_repository` now accepts the `strip_prefix` arg and passes it to the underlying `git_repository` call.

### CODE REVIEW

**Comment by Wyverald** [comment]
> * Add changelog entry
You can just include a line in the PR message that begins with `RELNOTES:`. The part after the `RELNOTES:` will be included in the release notes of the next release.

> * Figure out how to run tests and then add some.

You can add to/fix this one: https://cs.opensource.google/bazel/bazel/+/master:src/test/py/bazel/bzlmod/bazel_overrides_test.py;drc=3f9d80c35d88a280c2d53682f3d201b4733a3fff;l=234

To test locally, run `bazel test src/test/py/bazel:bazel_overrides_test --test_filter=<test_method>`.

File: src/test/py/bazel/bzlmod/bazel_overrides_test.py [hunk 1]
```diff
@@ -272,6 +275,101 @@ def testGitOverride(self):
...
+        cwd=src_aaa_1_0,
+    )
+    self.RunProgram(
+        ['git', 'config', 'user.email', 'tester@foo.com'],
+        cwd=src_aaa_1_0,
+    )
+
+    # Make a subdirectory that itself is the published module 'ddd'.
+    subdir_name = 'subdir_containing_ddd'
+    shutil.copytree(src=src_ddd_1_0, dst=src_aaa_1_0 / subdir_name)
```
Comment (dougthor42) [inline src/test/py/bazel/bzlmod/bazel_overrides_test.py:323]: I could also use `self.RunProgram(["cp", "-r", ...])` if that's better for some reason.

**Review by dougthor42 — COMMENTED** [review]
Thanks for the tip! That really sped things along - I don't know how long it would have taken me without it, haha.

Anyway, the code and PR message have updated. I also added a couple questions to the review. PTAL when you can.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java [hunk 2]
```diff
@@ -1015,10 +1015,18 @@ public void archiveOverride(
...
-            doc = "Whether submodules in the fetched repo should be recursively initialized.",
+            doc = "Whether git submodules in the fetched repo should be recursively initialized.",
             named = true,
             positional = false,
             defaultValue = "False"),
+        @Param(
+            name = "strip_prefix",
+            doc =
+                "A directory prefix to strip from the extracted files. This can be used to target"
+                    + " a subdirectory of the git repo.",
```
[the remark this answers is not in the document]
Comment (meteorcloudy) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java:1026]: Yes, that would be better!

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java [hunk 3]
```diff
@@ -1015,10 +1015,18 @@ public void archiveOverride(
             defaultValue = "0"),
         @Param(
             name = "init_submodules",
-            doc = "Whether submodules in the fetched repo should be recursively initialized.",
+            doc = "Whether git submodules in the fetched repo should be recursively initialized.",
```
[the remark this answers is not in the document]
Comment (meteorcloudy) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java:1018]: Thanks!

File: src/test/py/bazel/bzlmod/bazel_overrides_test.py (hunk 1 above)
Comment (meteorcloudy) [reply src/test/py/bazel/bzlmod/bazel_overrides_test.py:323]: No, that may not work on Windows.

File: src/test/py/bazel/bzlmod/bazel_overrides_test.py [hunk 4]
```diff
@@ -272,6 +275,101 @@ def testGitOverride(self):
...
+        str(src_aaa_relpath / subdir_name / "ddd.cc"),
+        [
+            '#include <stdio.h>',
+            '#include "ddd.h"',
+            'void hello_ddd(const std::string& caller) {',
+            '    std::string lib_name = "ddd@1.0";',
+            '    printf("%s => %s from subdir\\n", caller.c_str(), lib_name.c_str());',
+            '}'
+        ]
+    )
```
[the remark this answers is not in the document]
Comment (meteorcloudy) [reply src/test/py/bazel/bzlmod/bazel_overrides_test.py:338]: I think the current setup is fine.

**Review by meteorcloudy — APPROVED** [review]

**Pull request closed without merging**